text/0000-named-blocks.md

```diff
@@ -0,0 +1,635 @@
+  to pass them into another component (otherwise you'd have to use a
+  combinatoric mess of block syntax + `if hasBlock` checks to forward
+  blocks through to the inner component)
+
+I propose an opt-in API; any Ember Component that wants `main`/`else`
+properties to be set on the component instance need to opt into this
+behavior via something like:
+
+```
+const Component = Ember.Component.extend({
+  blockManager: inject.service(),
+  init() {
+    this._super();
+    this.get('blockManager').registerBlock(this.get('main'));
+  },
+});
+
+Component.reopenClass({
+  enableSingleBlockSyntaxProperties: true
+});
+
+export default Component;
+```
+
+So if `fancy-if` were an Ember component that opted-into
+`enableSingleBlockSyntaxProperties:true`, then
+given the component invocation:
+
+```
+{{#fancy-if cond=trueOrFalse}}
+  True
+{{else}}
+  False
+{{/fancy-component}}
+```
+
+The following ember component layout would work:
+
+```
+{{#if cond}}
+  {{main}}
+{{else}}
+  {{this.else}}
+{{/if}}
+```
+
+##### Alternatives? Nicer opt-in syntax?
+
+The `reopenClass` syntax s clunky, and I wouldn't be surprised if people
+wanted to use unified renderable syntax all over the place. So is there
+a less painful way to opt-in? Could it just be a property set on the
+prototype rather than the class object?
+
+# How We Teach This
+
+We teach this as a followup to classic block syntax; once the user is comfortable with single-block syntax, we can introduce named block syntax for more complex use cases.
+
+We teach that what `<@blockName></@blockName>` syntax really means is
+that we're just passing in an arg named `@blockName`, which is like
+any other arg we might pass into a component, but it happens to point
+to a template block than, say, some simple string value.
+
+# Drawbacks
+
+### Not your granddaddy's WC slot syntax
+
+This isn't really anything like the
+[WebComponents slot syntax](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/slot)
+that intends to address similar use cases, so there is some risk of
+introducing an API that doesn't fit in with what the rest of the world
+is doing.
+
+Some syntax highlighters might have trouble with this syntax; all
+the editors I've tried it on look reasonable, but GitHub's Handlebars
+parser isn't too kind:
+
+```hbs
+<x-modal>
+  <@header as |c|>
+    ...
+  </@header>
+
+  <@main as |c|>
+    ...
+  </@main>
+</x-modal>
+```
+
+### Conditionally passing blocks?
+
+This RFC does NOT introduce any kind facility for conditionally passing
+blocks, e.g.:
+
+```html
+{{! this syntax is INVALID! }}
+<x-layout>
+  <@header>...</@header>
+  <@main>...</@main>
+
+  {{#if userCanProceed}}
+    <@footer>
+      {{submit-button}}
+    </@footer>
+  {{/if}}
+</x-layout>
+```
+
+This might be desirable in the future, particularly for use cases
+involving flex-ish layouts where the component changes behavior /
+appearance based on whether blocks on passed in.
```

Comment:
Sure, that'd work, but the idea behind unified renderable syntax is that all you should really have to do is check if the arg/property is truthy or not; it's either present or it's not, and you shouldn't have to check for the presence of a specific type of renderable (be it a block or component factory), as that would arbitrarily force anyone invoking your component to use only specific kinds of renderables.

Then again... the only way to build components that support universal renderables is to commit to always passing args to universal renderables as both positional params and named k=v pairs, which, is kinda weird/bad?

Elsewhere in the RFC I demonstrated this as `{{header headerTitle title=headerTitle}}`, but maybe that's ridiculous to expect people to do, or maybe it only makes sense for addon authors to do, and app maintainers can just choose/prefer invocation styles that are consistent with the rest of the app? Will have to think on this...